### PR TITLE
client, cmd/hello, ipn, wgengine: identify netstack-forwarded clients…

### PR DESCRIPTION
**Pull request by namansood**
… in WhoIs

Updates #504

Updates #707

Signed-off-by: Naman Sood <mail@nsood.in>

### CODE REVIEW

File: ipn/ipnlocal/local.go [hunk 1]
```diff
@@ -253,14 +253,20 @@ func (b *LocalBackend) UpdateStatus(sb *ipnstate.StatusBuilder) {
 	}
 }
 
-// WhoIs reports the node and user who owns the node with the given IP.
+// WhoIs reports the node and user who owns the node with the given IP:port.
+// If the IP address is unique to this node, the port provided may be 0.
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:257]: If the IP address is a Tailscale IP, the provided port may be 0.

File: wgengine/netstack/netstack.go [hunk 2]
```diff
@@ -33,6 +33,7 @@ import (
 	"gvisor.dev/gvisor/pkg/tcpip/transport/udp"
 	"gvisor.dev/gvisor/pkg/waiter"
 	"inet.af/netaddr"
+	"tailscale.com/client/tailscale"
```
Comment (bradfitz) [inline wgengine/netstack/netstack.go:36]: we shouldn't use the client from the server. That's too indirect/slow/unreliable.

we should be able to do what it does directly, without going over HTTP.

File: wgengine/userspace.go [hunk 3]
```diff
@@ -1339,6 +1340,34 @@ func (e *userspaceEngine) Ping(ip netaddr.IP, cb func(*ipnstate.PingResult)) {
...
+func (e *userspaceEngine) UnregisterIPPortIdentity(ipport netaddr.IPPort) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	if e.nodeByIPPort == nil {
+		return
+	}
+	delete(e.nodeByIPPort, ipport)
+}
+
+func (e *userspaceEngine) WhoIsIPPort(ipport netaddr.IPPort) (*tailcfg.Node, bool) {
```
Comment (bradfitz) [inline wgengine/userspace.go:1361]: name result parameters: `n` and `ok`

File: wgengine/wgengine.go [hunk 4]
```diff
@@ -137,4 +137,15 @@ type Engine interface {
 	// Ping is a request to start a discovery ping with the peer handling
 	// the given IP and then call cb with its ping latency & method.
 	Ping(ip netaddr.IP, cb func(*ipnstate.PingResult))
+
+	// RegisterIPPortIdentity registers a given node as temporarily having
```
Comment (bradfitz) [inline wgengine/wgengine.go:141]: This need to say something about the ownership/mutation rules of the pointer value.

But I think we don't need the whole *Node. Just the userID, right?

File: ipn/ipnlocal/local.go [hunk 5]
```diff
@@ -253,14 +253,20 @@ func (b *LocalBackend) UpdateStatus(sb *ipnstate.StatusBuilder) {
...
-func (b *LocalBackend) WhoIs(ip netaddr.IP) (n *tailcfg.Node, u tailcfg.UserProfile, ok bool) {
+func (b *LocalBackend) WhoIs(ipport netaddr.IPPort) (n *tailcfg.Node, u tailcfg.UserProfile, ok bool) {
 	b.mu.Lock()
 	defer b.mu.Unlock()
-	n, ok = b.nodeByAddr[ip]
+	n, ok = b.nodeByAddr[ipport.IP]
 	if !ok {
-		return nil, u, false
+		if ipport.Port != 0 {
+			n, ok = b.e.WhoIsIPPort(ipport)
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:265]: yeah, looks like you only care about the userid. So let's make WhoIsIPPort return a tailcfg.UserID instead (a value type, not a pointer) so we don't need to worry about ownership/corruption/data races from potential mutations.

File: ipn/ipnlocal/local.go [hunk 6]
```diff
@@ -253,14 +253,20 @@ func (b *LocalBackend) UpdateStatus(sb *ipnstate.StatusBuilder) {
 	}
 }
 
-// WhoIs reports the node and user who owns the node with the given IP.
+// WhoIs reports the node and user who owns the node with the given IP:port.
+// If the IP address is unique to this node, the port provided may be 0.
 // If ok == true, n and u are valid.
-func (b *LocalBackend) WhoIs(ip netaddr.IP) (n *tailcfg.Node, u tailcfg.UserProfile, ok bool) {
+func (b *LocalBackend) WhoIs(ipport netaddr.IPPort) (n *tailcfg.Node, u tailcfg.UserProfile, ok bool) {
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:259]: ipp or ipPort or addr

File: ipn/ipnlocal/local.go (hunk 5 above)
Comment (namansood) [reply ipn/ipnlocal/local.go:265]: I think we also need the node itself? The userid is used to get `u` but this function returns both `u` and `n`, and `n` is a Node. is there a way to get that here from the userid?

Comment (bradfitz) [reply ipn/ipnlocal/local.go]: look 6 lines down

Comment (bradfitz) [reply ipn/ipnlocal/local.go]: oh, for the node, you can use nodeByAddr as the lookup... return the Tailscale IP instead?

File: wgengine/wgengine.go [hunk 7]
```diff
@@ -138,14 +138,14 @@ type Engine interface {
 	// the given IP and then call cb with its ping latency & method.
 	Ping(ip netaddr.IP, cb func(*ipnstate.PingResult))
 
-	// RegisterIPPortIdentity registers a given node as temporarily having
-	// the given IP:port for whois lookups.
-	RegisterIPPortIdentity(netaddr.IPPort, *tailcfg.Node)
+	// RegisterIPPortIdentity registers a given node (identified by its
+	// Tailscale IP) as temporarily having the given IP:port for whois lookups.
```
Comment (bradfitz) [inline wgengine/wgengine.go:142]: elaborate a bit more here to clarify that the IP is almost always 127.0.0.1 (or some other localhost address) and the port is some ephemeral port for proxying to localhost.

File: wgengine/userspace.go [hunk 8]
```diff
@@ -1341,32 +1341,32 @@ func (e *userspaceEngine) Ping(ip netaddr.IP, cb func(*ipnstate.PingResult)) {
...
+	delete(e.tsIPByIPPort, ipport)
 }
 
-func (e *userspaceEngine) WhoIsIPPort(ipport netaddr.IPPort) (n *tailcfg.Node, ok bool) {
+func (e *userspaceEngine) WhoIsIPPort(ipport netaddr.IPPort) (tsIP netaddr.IP, ok bool) {
 	e.mu.Lock()
 	defer e.mu.Unlock()
-	if e.nodeByIPPort == nil {
-		return nil, false
+	if e.tsIPByIPPort == nil {
```
Comment (bradfitz) [inline wgengine/userspace.go:1365]: you can delete these three lines.  it's valid to look things up in a nil map... it just returns zero value and false.

**Review by bradfitz — APPROVED** [review]
Looks good otherwise!

**Pull request merged**